torch/fx/experimental/symbolic_shapes.py

```diff
@@ -323,12 +323,28 @@ def constrain_range(a, *, min: Optional[int], max: Optional[int] = None):
     if not isinstance(a, SymInt):
         if not (min <= a <= max):
             raise ValueRangeError(f"Invalid value {a} for range [{min}:{max}]")
+
+        if (
+            (fake_mode := detect_fake_mode()) is not None and
+            getattr(fake_mode, "shape_env", None) is not None
+        ):
+            # If we are tracing with a fake mode then add this integer to the
+            # shape_env's var_to_range
+            sym_integer = sympy.Integer(a)
+            shape_env = fake_mode.shape_env
+            shape_env.var_to_range[sym_integer] = ValueRanges(min, max)
+            shape_env.var_to_stack[sym_integer] = ''.join(
```

Comment:
Is this guaranteed to be eventually used in a constraint violation? Or is it possible that we will populate the stack but not raise later?

Reason I ask is that formatting the trace is expensive, so we should only do it when required.

Reply:
Oh I see. Yeah when running eagerly, this would not result in a constraint violation. Would just storing `TracingContext.extract_stack()` here and formatting it later in dynamo.export be better?

Reply:
Yeah that should be OK.

Reply:
Hmm, apologies for the churn. It seems that we've now broken the value type of `var_to_stack`. Sometimes it's a stack, sometimes it's a formatted string. :(

Reply:
OK so @ezyang thinks we can defer formatting for other sites that update `var_to_stack` as well. In which case please fix them up along with the declared type.

Reply:
updated!